sklearn/ensemble/_voting.py

```diff
@@ -252,6 +252,26 @@ def __init__(self, estimators, *, voting='hard', weights=None,
         self.flatten_transform = flatten_transform
         self.verbose = verbose
 
+    def _validate_estimators(self):
+        names, clfs = super()._validate_estimators()
+        # mapping from original labels to encoded ones
+        label_mapping = {c: i for i, c in enumerate(self.classes_)}
+
+        updated_clfs = tuple()
+        for clf in clfs:
+            if clf == 'drop':
+                continue
+            is_clone = False
+            for k, v in clf.get_params(deep=True).items():
+                if k.endswith('class_weight') and v is not None:
```

Comment:
`v` can be a string. For example, `LogisticRegression`'s class_weight can be `'balanced'`. We would need guard against it. I also prefer to use the following style so we have less indented lines:

```python
if not k.endswith('class_weight') or not isinstance(v, Mapping):
    continue

if not already_cloned:
    clf = clone(clf)
    already_cloned = True
....
```

Reply:
Hi Thomas, I've updated the code to check against the string argument values.

Reply:
> I am trying to think of ways where lead to unexpected or surprising behavior. There is a weird case where a third party estimator can name a parameter `class_weight` that has a completely different meaning than us. In that case, the automatically mapping may break their code.

Hi Thomas, I'm thinking maybe we could check if the estimator is a third party one, and if so, we could raise a warning if they have a parameter `class_weight`. WDYT?

Reply:
Looking at this again, we clearly define [class_weights](https://scikit-learn.org/stable/glossary.html#term-class_weight) in the glossary so we can assume third party estimators use the same definition.

The underlying issue with passing information from a metaestimator to inner estimators is very common in sklearn. In this case, we are trying to use the classes information from the metaestimator to re-encode the `class_weights` in the inner estimator.

I am unable to think of a good alternative to this PR without adding more API. I'm thinking broadly to see if we want to apply the same approach to other meta-estimators that have the same issue, (`StackingClassifier`).